soroban-spec/src/gen/json.rs

```diff
@@ -0,0 +1,283 @@
+use serde::Serialize;
+use soroban_env_host::xdr::{
+    Error, ScSpecEntry, ScSpecFunctionInputV0, ScSpecTypeDef, ScSpecUdtStructFieldV0,
+    ScSpecUdtUnionCaseV0,
+};
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct StructField {
+    name: String,
+    value: Type,
+}
+
+impl TryFrom<&ScSpecUdtStructFieldV0> for StructField {
+    type Error = Error;
+
+    fn try_from(f: &ScSpecUdtStructFieldV0) -> Result<Self, Self::Error> {
+        Ok(StructField {
+            name: f.name.to_string()?,
+            value: Type::try_from(&f.type_)?,
+        })
+    }
+}
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct FunctionInput {
+    name: String,
+    value: Type,
+}
+
+impl TryFrom<&ScSpecFunctionInputV0> for FunctionInput {
+    type Error = Error;
+
+    fn try_from(f: &ScSpecFunctionInputV0) -> Result<Self, Self::Error> {
+        Ok(FunctionInput {
+            name: f.name.to_string()?,
+            value: Type::try_from(&f.type_)?,
+        })
+    }
+}
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct UnionCase {
+    name: String,
+    values: Vec<Type>,
+}
+
+impl TryFrom<&ScSpecUdtUnionCaseV0> for UnionCase {
+    type Error = Error;
+
+    fn try_from(c: &ScSpecUdtUnionCaseV0) -> Result<Self, Self::Error> {
+        Ok(UnionCase {
+            name: c.name.to_string()?,
+            values: c
+                .type_
+                .as_ref()
+                .map(Type::try_from)
+                .transpose()?
+                .into_iter()
+                .collect(),
+        })
+    }
+}
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize)]
+#[serde(tag = "type")]
+#[serde(rename_all = "camelCase")]
+pub enum Type {
+    U64,
+    I64,
+    U32,
+    I32,
+    Bool,
+    Symbol,
+    Bitset,
+    Status,
+    Bytes,
+    BigInt,
+    Map { key: Box<Type>, value: Box<Type> },
+    Option { value: Box<Type> },
+    Result { value: Box<Type>, error: Box<Type> },
+    Set { element: Box<Type> },
+    Vec { element: Box<Type> },
+    Tuple { elements: Vec<Type> },
+    Custom { name: String },
+}
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize)]
+#[serde(tag = "type")]
+#[serde(rename_all = "camelCase")]
+pub enum Entry {
+    Function {
+        name: String,
+        inputs: Vec<FunctionInput>,
+        outputs: Vec<Type>,
+    },
+    Struct {
+        name: String,
+        fields: Vec<StructField>,
+    },
+    Union {
+        name: String,
+        cases: Vec<UnionCase>,
+    },
+}
```

Comment:
@tamirms I was about to merge but noticed the entries were bundled with the list of types. I split them out separately because I think they're two different things, it doesn't have any material impact on the JSON though.

Reply:
thanks @leighmcculloch I agree that it's much cleaner with your modifications!